### ISSUE TITLE
How to interact with Electron app controls using Spectron.

### ISSUE THREAD

**Issue by Bharadwaj108** [electron/spectron#143]
Hello,
My project is using Electron framework. The developers have made a built and have given me an electron exe. When you click and open the exe it displays a login frame which is the entry point of my application. The form basically has 3 controls,user name,password text boxes and Sign in button. Now I have to perform a functional test on that page. My requirement is Input some text into the user name and password text box and click on the submit button. (in this case the assumption is a successful login). Once logged in I see a welcome message which I have to test. My question is how do you get the textbox and submit button references from the electron app and perform actions on it.  I have tried the below code,it open my electron exe but does nothing and test case passes.
```
var assert = require('assert');
var Application = require('spectron').Application
describe('Test Suite 1', function () {
    this.timeout(100000)
    beforeEach(function () {
        this.app = new Application({
            path: '/CorumDispense-win32-ia32/CorumDispense.exe'
        })

        return this.app.start()        
    })
    /*afterEach(function () {
        if (this.app && this.app.isRunning()) {
            return this.app.stop()            
        }
    })*/
    it('Login Dispense', function () {
        var txtUserName = this.app.client.elementIdText('inputUsername3');
        var txtPassword = this.app.client.elementIdText('inputPassword3');   
        var btnSignIn = this.app.client.element('//button/*[text(),Sign in]');
        txtUserName.keys('test user');     
        txtPassword.keys('test123');     

        //click on signin button
       btnSignIn.click();

    })
})
```
The output is the test case has passed but I dont see any text being keyed into the user name or the password and the button has not been clicked

```
Test Name:	Test Suite 1 Login Dispense
Test Outcome:	Passed
Result StandardOutput:	
mocha.json options file not found. Using default values: { ui: 'tdd', reporter: 'tap', timeout: 2000 }
1..1
ok 1 Test Suite 1 Login Dispense
# tests 1
# pass 1
# fail 0



```

**My package.json**

```
{
  "name": "lighthouse",
  "version": "0.0.0",
  "private": true,
  "scripts": {
    "start": "node ./bin/www"
  },
  "description": "Lighthouse",
  "author": {
    "name": "bpappu"
  },
  "dependencies": {
    "body-parser": "~1.8.1",
    "cookie-parser": "~1.3.3",
    "debug": "~2.0.0",
    "express": "~4.9.0",
    "jade": "~1.6.0",
    "morgan": "~1.3.0",
    "serve-favicon": "~2.1.3",
    "spectron": "^3.4.0",
    "stylus": "0.42.3"
  },
  "devDependencies": {
    "chai": "^3.5.0",
    "electron": "^1.4.6",
    "electroner": "^4.0.2",
    "mocha": "^3.1.2",
    "spectron": "^3.4.0",
    "spectron-keys": "0.0.1"
  }
}
```

Can some one please tell me if I am missing anything here. 
cheers,
Bharadwaj.

**Comment by Bharadwaj108** [comment]
Hello guys,Can  I please get some help here. I am stuck with this electron app testing for quiet some time now with out much success.Can any one here provide me with a solution for my above problem.

**Comment by pmoleri** [comment]
@Bharadwaj108 keep in mind that WebdriverIO is async, and your `it` method is sync, that's why the test case may pass even before any code is actually executed in electron.
Take a look to mocha on how to [test asynchronous code](https://mochajs.org/#asynchronous-code), in particular to the Promises section because WebdriverIO returns promises.

For your code to work, you need to chain your promises and return them.
i.e.
```js
   it('Login Dispense', function () {
        return
            this.app.client.elementIdText('inputUsername3')
                .then(response => {
                    response.value.keys('test user');    // some kind of assert?
                    return this.app.client.elementIdText('inputPassword3');
                }).then(response => {
                    response.value.keys('test123');
                    return this.app.client.element('//button/*[text(),Sign in]').click();
                });
    })
```

**Comment by kevinsawicki** [comment]
Closing this out, thanks for the example @pmoleri.

@Bharadwaj108 See http://webdriver.io/api.html for more details and examples.

**Comment by Bharadwaj108** [comment]
Hello guys ,
I am very sorry for the delayed response. I got a chance to solve this issue(but I am now using Selenium ChromeDriver  and C# as my QA  Technology). The reason why I could not interact with the controls was when my electron application was opened by using my above code,the developer tools window was enabled when the application was complied and was released for QA. That window actually conflicted with the actual application and thus interrupted with me accessing the controls. As soon as I recompiled the application with the developer tools window disabled I could work with the controls inside my Electron app.
Thanks a lot for the positive response guys appreciate it.

Regards,
bharadwaj.
